Tutorial on custom mutations (#1746)
keystonejs/keystone-5: docs/guides/custom-mutations.md

<!--[meta]
section: tutorials
title: Custom Mutations
[meta]-->

# Custom Mutations

Out of the box Keystone provides predictable CRUD operations (Create, Read, Update and Delete) for Lists. The generated graphQL queries and mutations are the primary method for updating data in a List. These should be enough for most application requirements.

Custom Queries and Mutations may be required if you wish to preform non-CRUD operations or actions that don't relate to a specific List.

See the [GraphQL Philosophy](https://v5.keystonejs.com/guides/graphql-philosophy) for more information on how Keystone implements CRUD operations in GraphQL and when Custom Queries and Mutations may be required.

> You can add to Keystone's generated schema with custom types, queries, and mutations using the `keystone.extendGraphQLSchema()` method.\

Reviewer: Wayward `\`.

> You can add to Keystone's generated schema with custom types, queries, and mutations using the `keystone.extendGraphQLSchema()` method.\

Author: ```suggestion
You can add to Keystone's generated schema with custom types, queries, and mutations using the `keystone.extendGraphQLSchema()` method.
```

# Creating an Custom Mutation

A common example where a custom mutation might be beneficial is if you want to increment a value.

> Like any problem there are multiple solutions. You can implement an incrementing value with (Hooks)[https://v5.keystonejs.com/guides/hooks] but in this example we're going to look at how to do this with a custom mutation.

Reviewer: `()` and `[]` swapped in link.

> Like any problem there are multiple solutions. You can implement an incrementing value with (Hooks)[https://v5.keystonejs.com/guides/hooks] but in this example we're going to look at how to do this with a custom mutation.

Author: ```suggestion
Like any problem there are multiple solutions. You can implement an incrementing value with [Hooks](https://v5.keystonejs.com/guides/hooks) but in this example we're going to look at how to do this with a custom mutation.
```

Each item in the mutation array requires a `schema` and a `resolver`.

The Schema defines the input and return types of the mutation. You can [learn more about schemas and types](https://graphql.org/learn/schema/) on [https://graphql.org](https://graphql.org).

Our Schema is simple: `incrementPageViews(id: ID!): Page`. It's called `incrementPageViews` and requires an `id` parameter that must be of an `ID` type. It returns a `Page` type.

`ID` is an internal GraphQL type and `Page` is a type that has been generated automatically by Keystone when we created the `Page` list.

The resolver is a function that returns an object matching the `Page` type. You can see all the type generated by Keystone by looking at the GraphQL Playground documentation that is accessible via the admin UI.

All together our custom mutation looks like this:

```js
keystone.extendGraphQLSchema({
  mutations: [
    {
      schema: `incrementPageViews(id: ID!): Page`,
      resolver: incrementPageViews,
    },
  ],
});
```

> The last step is to define the resolver function `incrementPageViews`.

Reviewer: Is this sentence meant to go a couple of paragraphs down?

> The last step is to define the resolver function `incrementPageViews`.

Author: ```suggestion
```

In this mutation we want to access an existing item in the list. We don't care about access control, in-fact we will make the field read-only so that it cannot be updated with normal mutations or in the AdminUI. Our new Page definition is:

```js
const Page = {
  fields: {
    title: { type: Text },
    views: {
      type: Integer,
      access: {
        create: false,
        read: true,
        update: false,
      },
    },
  },
};
```

> 

Author: ```suggestion

The last step is to define the resolver function `incrementPageViews`.

```

Our function will by-pass access control and update the value directly by getting the list item and then by calling `findById` on the list adapter. Once we have the old item we can call `update` with the new values.

```js
const incrementPageViews = async (_, { id }) => {
  const list = keystone.lists.Page;
  const oldItem = await list.adapter.findById(id);
  const newItem = await list.adapter.update(id, {
    ...oldItem,
    views: (oldItem.views || 0) + 1,
  });
  return newItem;
};
```

Note: The value of `views` may be `undefined` initially, so before we increment it we make sure to convert any `falsey` values to `0`.

> Our custom mutation is now available to the client that trigger it with the following graphQL query:

Reviewer: Is this sentence missing a word or something?

> Our custom mutation is now available to the client that trigger it with the following graphQL query:

Author: ```suggestion
Our custom mutation is now available to the client that can use it to increment page views like this:
```

```
mutation incrementPageViews($id: ID!) {
  mutation {
    incrementPageViews(id: $id) {
      views

>   }

Author: views is an integer
